src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java

```diff
@@ -50,19 +50,58 @@ public final class ObjectFactoriesFilter {
     /**
      * Checks if serial filter configured with {@code "jdk.jndi.object.factoriesFilter"}
      * system property value allows instantiation of the specified objects factory class.
-     * If the filter result is not {@linkplain Status#REJECTED REJECTED}, the filter will
+     * If the filter result is {@linkplain Status#ALLOWED ALLOWED}, the filter will
      * allow the instantiation of objects factory class.
      *
-     * @param factoryClass objects factory class
+     * @param clz objects factory class
      * @return true - if the factory is allowed to be instantiated; false - otherwise
      */
-    public static boolean canInstantiateObjectsFactory(Class<?> factoryClass) {
-        return checkInput(() -> factoryClass);
+    public static boolean checkGlobalFilter(Class<?> clz) {
+        return checkInput("global", () -> clz);
     }
 
-    private static boolean checkInput(FactoryInfo factoryInfo) {
-        Status result = GLOBAL.checkInput(factoryInfo);
-        return result != Status.REJECTED;
+    /**
+     * Checks if the factory filters allow the given factory class for LDAP.
+     * This method combines the global and LDAP specific filter results to determine
+     * if the given factory class is allowed.
+     * The given factory class is rejected if any of these two filters reject
+     * it, or if none of them allow it.
+     *
+     * @param clz objects factory class
+     * @return true - if the factory is allowed to be instantiated; false - otherwise
+     */
+    public static boolean checkLdapFilter(Class<?> clz) {
+        return checkInput("ldap", () -> clz);
+    }
+
+    /**
+     * Checks if the factory filters allow the given factory class for RMI.
+     * This method combines the global and RMI specific filter results to determine
+     * if the given factory class is allowed.
+     * The given factory class is rejected if any of these two filters reject
+     * it, or if none of them allow it.
+     *
+     * @param clz objects factory class
+     * @return true - if the factory is allowed to be instantiated; false - otherwise
+     */
+    public static boolean checkRmiFilter(Class<?> clz) {
+        return checkInput("rmi", () -> clz);
+    }
+
+    private static boolean checkInput(String scheme, FactoryInfo factoryInfo) {
+        Status result = switch(scheme) {
```

Comment:
The handling of the selection of the filter could be more direct.
You can change the argument to checkInput to be ObjectInputFilter and pass the desired filter; LDAP_FILTER, RMI_FITLER, or GLOBAL_FILTER.
And make the check of the GLOBAL_FILTER conditional on it not having already been evaluated.
Then it will be the case that there must always be a specific filter.

The callers are all local to the class and would change to pass the desired filter.

Reply:
Thank you - refactored as suggested.